felix/ipsets/ipsets.go

```diff
@@ -124,109 +150,116 @@ func (s *IPSets) AddOrReplaceIPSet(setMetadata IPSetMetadata, members []string)
-	canonMembers := s.filterAndCanonicaliseMembers(setMetadata.Type, members)
 
-	// Create the IP set struct and store it off.
-	setID := setMetadata.SetID
-	ipSet := &ipSet{
-		IPSetMetadata:    setMetadata,
-		MainIPSetName:    s.IPVersionConfig.NameForMainIPSet(setID),
-		pendingReplace:   canonMembers,
-		pendingAdds:      set.New[IPSetMember](),
-		pendingDeletions: set.New[IPSetMember](),
+	// Mark that we want this IP set to exist and with the correct size etc.
+	// If the IP set exists, but it has the wrong metadata then the
+	// DeltaTracker will catch that and mark it for recreation.
+	mainIPSetName := s.IPVersionConfig.NameForMainIPSet(setID)
+	dpMeta := dataplaneMetadata{
+		Type:    setMetadata.Type,
+		MaxSize: setMetadata.MaxSize,
+	}
+	s.setNameToAllMetadata[mainIPSetName] = dpMeta
+	if s.ipSetNeeded(mainIPSetName) {
+		s.setNameToProgrammedMetadata.Desired().Set(mainIPSetName, dpMeta)
 	}
-	s.ipSetIDToIPSet[setID] = ipSet
-	s.mainIPSetNameToIPSet[ipSet.MainIPSetName] = ipSet
 
-	// Mark IP set dirty so ApplyUpdates() will rewrite it.
-	s.dirtyIPSetIDs.Add(setID)
+	// Set the desired contents of the IP set.
+	canonMembers := s.filterAndCanonicaliseMembers(setMetadata.Type, members)
+	memberTracker := s.getOrCreateMemberTracker(mainIPSetName)
 
-	// The IP set may have been previously queued for deletion, undo that.
-	s.pendingIPSetDeletions.Discard(ipSet.MainIPSetName)
+	desiredMembers := memberTracker.Desired()
+	desiredMembers.Iter(func(k IPSetMember) {
+		if canonMembers.Contains(k) {
+			canonMembers.Discard(k)
+		} else {
+			desiredMembers.Delete(k)
+		}
+	})
+	canonMembers.Iter(func(m IPSetMember) error {
+		desiredMembers.Add(m)
+		return nil
+	})
+	s.updateDirtiness(mainIPSetName)
+}
+
+func (s *IPSets) getOrCreateMemberTracker(mainIPSetName string) *deltatracker.SetDeltaTracker[IPSetMember] {
+	dt := s.mainSetNameToMembers[mainIPSetName]
+	if dt == nil {
+		dt = deltatracker.NewSetDeltaTracker[IPSetMember]()
+		s.mainSetNameToMembers[mainIPSetName] = dt
+	}
+	return dt
 }
 
 // RemoveIPSet queues up the removal of an IP set, it need not be empty.  The IP sets will be
 // removed on the next call to ApplyDeletions().
 func (s *IPSets) RemoveIPSet(setID string) {
 	s.logCxt.WithField("setID", setID).Info("Queueing IP set for removal")
-	delete(s.ipSetIDToIPSet, setID)
-	mainIPSetName := s.IPVersionConfig.NameForMainIPSet(setID)
-	delete(s.mainIPSetNameToIPSet, mainIPSetName)
-	s.dirtyIPSetIDs.Discard(setID)
-	s.pendingIPSetDeletions.Add(mainIPSetName)
+	// Mark that we no longer need this IP set.  The DeltaTracker will keep track of the metadata
+	// until we actually delete the IP set.  We clean up mainSetNameToMembers only when we actually
+	// delete it.
+	setName := s.nameForMainIPSet(setID)
+	delete(s.setNameToAllMetadata, setName)
+	s.setNameToProgrammedMetadata.Desired().Delete(setName)
+	if _, ok := s.setNameToProgrammedMetadata.Dataplane().Get(setName); ok {
+		s.mainSetNameToMembers[setName].Desired().DeleteAll()
```

Comment:
when will this get removed from `mainSetNameToMembers`?  This looks to me like it only does "delete all members from the set"

Reply:
It happens in ApplyDeletions around line 843.  Basically we keep track of the members until we actually delete the IP set from the dataplane.  It can also happen via a resync.